packages/core-contracts/contracts/utils/SafeCast.sol

```diff
@@ -2,70 +2,92 @@
 pragma solidity ^0.8.0;
 
 /**
- * @title TODO Utility that avoids silent overflow errors.
+ * @title Utility that avoids silent overflow errors.
  *
  * Example silent overflow errors in Solidity:
  * 1) If uint256 A = type(uint128).max + 1, down casting A to uint128 returns zero.
  * 2) If int256 A = -1, casting A to uint256 returns type(uint256).max.
+ *
+ * Visual helper:
+ * uint256 [0, 2^256 - 1]          -------------------------------o===============================>
+ * int256  [-2^255 - 1, 2^255 - 1] ----<==========================o===========================>----
+ * uint128 [0, 2^128 - 1]          -------------------------------o===============>----------------
+ * int128 [-2^127 - 1, 2^127 - 1]  ----------------<==============o==============>-----------------
  */
 library SafeCast {
     error CastError(bytes32 fromType, bytes32 toType);
 
     // Note: Overloading doesn't seem to work for similar types, i.e. int256 and int128, uint256 and uint128, etc, so explicitly naming the functions differently here.
 
     function uint256toUint128(uint256 x) internal pure returns (uint128) {
+        // -------------------------------o===============================>
+        // -------------------------------o===============>xxxxxxxxxxxxxxxx
         if (x > type(uint128).max) {
-            revert("Failed cast uint256 to uint128");
+            revert CastError("uint256", "uint128");
```

Comment:
these constants are only better assuming solidity is able to optimize the strings down from repetitions, but that seems doubtful. It might be better to just create one error per failure, ie

```
error InvalidU256ToU128(uint256 invalid);
```

Reply:
Cool. Implemented.

Reply:
wohoo!